docs/src/adr/005adr-json.md

```diff
@@ -0,0 +1,121 @@
+# ADR-005: JSON Serialization Format
+
+| author       | revision |
+| ------------ | --------:|
+| Jure Kukovec |        1 |
+
+This ADR documents our decision on serializing the Apalache internal representation (IR) as JSON.
+The purpose of introducing such a serialization is to expose the internal representation in a standardized format, which can be used for persistent storage, or for analysis by third-party tools in the future.
+
+## 1. Serializable classes
+
+The following classes are serializable:
+1. TLA+ expressions `TlaEx` and subclasses thereof:
+  1. Named expressions `NameEx`
+  1. Literal values `ValEx` for the following literals:
+    1. Integers `TlaInt`
+    1. Strings `TlaStr`
+    1. Booleans `TlaBool`
+    1. Decimals `TlaDecimal`
+  1. Operator expressions `OperEx`
+  1. LET-IN expressions `LetInEx`
+1. TLA+ declarations `TlaDecl` and subclasses thereof:
+  1. Variable declarations `TlaVarDecl`
+  1. Constant declarations `TlaConstDecl`
+  1. Operator declarations `TlaOperDecl`
+  1. Assumption declarations `TlaAssumeDecl`
+  1. Theorem declarations `TlaTheoremDecl`
+1. TLA+ modules `TlaModule`
+
+## 2. Disambiguation field
+
+Every serialization will contain a disambiguation field, named `kind`. This field holds the name of the class being serialized. For example, the serialization of a `NameEx` will have the shape
+```
+{
+  kind: "NameEx"
+  ...
+}
+```
+
+## 3. Serializing typed entities
+
+Serializations of typed entities will have an additional field named `type`, containing the type of the expression (see [ADR-002][], [ADR-004][] for a description of our type system and the syntax for types-as-string-annotations respectively). For example, the integer literal `1` is represented by a `ValEx`, which has type `Int` and is serialized as follows:
+```
+{
+  kind: "NameEx"
+  type: "Int"
+  ...
+}
+```
+
+## 4. Serialization rules
```

Comment:
Can you add examples for all kinds of declarations? This should not be hard. This will help the reader to see the format without running the tool. Perhaps, you would want to use the simplified format for the parameters, which I mentioned in a comment to this PR.

Reply:
That kind of documentation will surely be useful for readers, but I don't think manually generating it for the ADR is a very good use of time.

Once it's implemented, we can just automatically generate it for the docs (and even ensure that the docs stay up to date).

Reply:
That would be even better